libnetwork, libn/d/overlay: handle coalesced updates from NetworkDB

corhere opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
- Fixed #49908 
- Fixed #48142 

**- How I did it**
The eventually-consistent nature of NetworkDB means we cannot depend on events being received in the same order that they were sent. Nor can we depend on receiving events for all intermediate states. It is possible for a series of entry UPDATEs, or a DELETE followed by a CREATE with the same key, to get coalesced into a single UPDATE event on the receiving node. Watchers of NetworkDB tables therefore need to be prepared to gracefully handle arbitrary UPDATEs of a key, including those where the new value may have nothing in common with the previous value.

Both the overlay driver's watcher for `overlay_peer_table` and the libnetwork controller's watcher for `endpoint_table` would misbehave when processing a coalesced UPDATE event. They would handle it the same as a CREATE: insert new entries into control-plane databases without removing the old. (This is a simplified explanation.) The result would be "transient states" with conflicts that never resolve, leaving obsolete entries in kernel control-plane databases and service-discovery DNS. Fix the watchers to remove the old before adding the new when handling UPDATE events.

**- How to verify it**
1. Set up a three-node Swarm with two workers and one manager
2. Create a user-defined, attachable overlay network
3. On the worker nodes, set up a simulated lossy underlay network using the `tc netem` scheduler. E.g.: `tc qdisc add dev eth0 root netem delay 100ms 200ms loss 1%`
4. Verify that service-discovery DNS stays in sync:
    1. Deploy a globally-replicated service with constraint `node.role!=manager` which is connected to the user-defined overlay network from step 2. Arrange it so the tasks exit after a few seconds, e.g. `alpine sh -c 'sleep 5'`
    2. Run a basic container on the manager node, connected to the overlay network from step 2. Repeatedly make DNS queries for `tasks.<service from step 4>` while the service tasks keep flapping in the background. Verify that the set of IP addresses in the answer does not grow unbounded.
    3. Verify that the the daemon logs on all three nodes show no evidence of persistent transient states
    4. Delete the basic container and service
5. Verify that service-discovery DNS and overlay FDB do not fall out of sync when the same endpoint ID is repeatedly removed and re-added:
    1. On the manager node, `docker run -d --name hellobasic --net my-overlay nginxdemos/hello:plain-text`
    2. On one of the worker nodes, `docker run --rm --net my-overlay alpine watch wget -qO- hellobasic`
    3. Back on the manager, repeatedly `for run in {1..3}; do docker kill hellobasic && docker start hellobasic; done` so the endpoint flaps and gets reattached with different IP and MAC addresses.
    4. Verify that the worker node's container is eventually able to communicate with the hellobasic container after each restart.
    5. Delete the containers
6. Verify that the ingress routing mesh is reliable with flapping tasks:
    1. `docker service create -d --name hello --network my-overlay --mode global --constraint 'node.role!=manager' -p 30888:80 nginxdemos/hello:plain-text`
    2. curl port 30888 on each node, repeatedly. Verify that repeated requests all succeed, and hit different tasks.
    3. `for run in {1..10}; do docker service update --force hello; sleep 15; done`
    4. Repeat step ii.

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
- Greatly improve the reliability of overlay networking and the Swarm routing mesh
```

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
libn/networkdb: report prev value in update events

When handling updates to existing entries, it is often necessary to know
what the previous value was. NetworkDB knows the previous and new values
when it broadcasts an update event for an entry. Include both values in
the update event so the watchers do not have to do their own parallel
bookkeeping.

Unify the event types under WatchEvent as representing the operation kind
in the type system has been inconvenient, not useful. The operation is
now implied by the nilness of the Value and Prev event fields.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libnetwork/driverapi: make EventNotify optional

Overlay is the only driver which makes use of the EventNotify facility,
yet all other driver implementations are forced to provide a stub
implementation. Move the EventNotify and DecodeTableEntry methods into a
new optional TableWatcher interface and remove the stubs from all the
other drivers.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libnetwork/d/overlay: handle coalesced peer updates

The eventually-consistent nature of NetworkDB means we cannot depend on
events being received in the same order that they were sent. Nor can we
depend on receiving events for all intermediate states. It is possible
for a series of entry UPDATEs, or a DELETE followed by a CREATE with the
same key, to get coalesced into a single UPDATE event on the receiving
node. Watchers of NetworkDB tables therefore need to be prepared to
gracefully handle arbitrary UPDATEs of a key, including those where the
new value may have nothing in common with the previous value.

The overlay driver naively handled events for overlay_peer_table
assuming that an endpoint leave followed by a rejoin of the same
endpoint would always be expressed as a DELETE event followed by a
CREATE. It would handle a coalesced UPDATE as a CREATE, inserting a new
entry into peerDB without removing the old one. This would
have various side effects, such as having the "transient state" of
multiple entries in peerDB with the same peer IP never settle.

Update driverapi to pass both the previous and new value of a table
entry into the driver. Modify the overlay driver to handle an UPDATE by
removing the previous peer entry from peerDB then adding the new one.
Modify the Windows overlay driver to match.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libnetwork: handle coalesced endpoint events

The eventually-consistent nature of NetworkDB means we cannot depend on
events being received in the same order that they were sent. Nor can we
depend on receiving events for all intermediate states. It is possible
for a series of entry UPDATEs, or a DELETE followed by a CREATE with the
same key, to get coalesced into a single UPDATE event on the receiving
node. Watchers of NetworkDB tables therefore need to be prepared to
gracefully handle arbitrary UPDATEs of a key, including those where the
new value may have nothing in common with the previous value.

The libnetwork controller naively handled events for endpoint_table
assuming that an endpoint leave followed by a rejoin of the same
endpoint would always be expressed as a DELETE event followed by a
CREATE. It would handle a coalesced UPDATE as a CREATE, adding a new
service binding without removing the old one. This would
have various side effects, such as having the "transient state" of
having multiple conflicting service bindings where more than one
endpoint is assigned an IP address never settling.

Modify the libnetwork controller to handle an UPDATE by removing the
previous service binding then adding the new one.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented:
Looks like there's a compile failure on Windows at least;

```
42.33 # github.com/docker/docker/libnetwork/drivers/windows
42.33 libnetwork/drivers/windows/windows.go:308:46: undefined: driverapi.EventType
60.13 + rm -f /go/src/github.com/docker/docker/go.mod

corhere committed:
libn/d/overlay: extract hashable address types

The macAddr and ipmac types are generally useful within libnetwork. Move
them to a dedicated package and overhaul the API to be more like that of
the net/netip package.

Update the overlay driver to utilize these types, adapting to the new
API.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libn/d/win/overlay: dedupe NetworkDB definitions

Windows and Linux overlay driver instances are interoperable, working
from the same NetworkDB table for peer discovery. As both drivers
produce and consume serialized data through the table, they both need to
have a shared understanding of the shape and semantics of that data.
The Windows overlay driver contains a duplicate copy of the protobuf
definitions used for marshaling and unmarshaling the NetworkDB peer
entries for dubious reasons. It gives us the flexibility to have the
definitions diverge, which is only really useful for shooting ourselves
in the foot.

Make daemon/libnetwork/drivers/overlay the source of truth for the peer
record definitions and the name of the NetworkDB table for distributing
peer records.

Signed-off-by: Cory Snider <csnider@mirantis.com>

robmry commented on daemon/libnetwork/agent.go:
```diff
@@ -909,21 +882,19 @@ func (c *Controller) handleEpTableEvent(ev events.Event) {
-	switch ev.(type) {
-	case networkdb.CreateEvent, networkdb.UpdateEvent:
+	switch {
+	case event.Value != nil:
```
Using the event.IsX methods might make this clearer?

corhere replied:
Yes, in retrospect I agree. I'll save this sort of construct for the following commit where the implementation gets corrected.

robmry reviewed (approved):
LGTM

austinvazquez reviewed (approved):
The networking stack is new to me, but 👍🏼  from a Go/delta POV. No blockers from me.

akerouanton reviewed (commented):
In your 2nd commit, introducing `TableWatcher`, you added an interface assertion to the Linux overlay driver, but not to the Windows overlay driver. Is this on purpose?

Otherwise LGTM
